src/main/java/at/ac/tuwien/kr/alpha/common/Program.java

```diff
@@ -0,0 +1,35 @@
+package at.ac.tuwien.kr.alpha.common;
+
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Alpha-internal representation of an ASP program, i.e., a set of ASP rules.
+ * Copyright (c) 2017, the Alpha Team.
+ */
+public class Program {
+	public static final Program EMPTY = new Program(Collections.emptyList(), Collections.emptyList());
+
+	private final List<Rule> rules;
+	private final List<Atom> facts;
+
+	public Program(List<Rule> rules, List<Atom> facts) {
+		this.rules = rules;
```

Comment:
IMO, the private members of `Program` should be `final`, with the consequence of a `Program` being immutable. This can be achieved by wrapping the arguments with [`Collections.unmodifiableList`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#unmodifiableList-java.util.List-).

This would force anyone to first come up with a set of rules (and facts) and only then instantiate a program, which then can never be changed. It makes some things simpler for us, e.g. no one can add a rule to the program while we are currently solving it (which might otherwise lead to undefined behavior).

In the `ParsedTreeVisitor` you'd probably have to accumulate two separate lists, and construct the `Program` from them.

If you that's too cumbersome, I am happy to make these changes in a follow-up PR.

Reply:
I thought about the same problem but decided against making `Program` immutable for the following reasons:
1. Each program transformation would have to create copies (I have plans to add two or three in the next days/weeks). We may have hundreds of thousands of facts, potentially also many rules. Copying seems very unnecessary when just a few rules get added and removed.
2. At Aalto, we have been thinking about lazy instantiating some transformations, e.g., to introduce aggregates. In a ground-and-solve setting these aggregates have to be instantiated during grounding, but in our setting we can postpone that (possibly even until very late in the solving). Not fully sure yet, but in these cases it might be necessary to have a mutable `Program`.
3. If we introduce an outward-facing API, we could just deep-copy the `Program` given to the API. Then a later modification from the outside does not touch the solver's private copy, hence no undefined behaviour.

Maybe 3. is not the best solution to this problem, so I am open to other approaches.

Reply:
OK, I agree :)